File: calculators/src/sumup_calculator.rs
use super::{Calculator, Charge, Material, Sale};
use super::{ChargeAmount, Config, SaleBreakdown};

const TRANSACTION_FEE_CARD_READER_NO_CONTRACT: f64 = 0.0169;
const TRANSACTION_FEE_POS_LITE_NO_CONTRACT: f64 = 0.0169;
const TRANSACTION_FEE_TAP_TO_PAY_IPHONE_NO_CONTRACT: f64 = 0.0169;
const TRANSACTION_FEE_REMOTE_PAYMENT_NO_CONTRACT: f64 = 0.025;
const TRANSACTION_FEE_QR_CODE_NO_CONTRACT: f64 = 0.0;
const TRANSACTION_FEE_CARD_READER_SUMUP_ONE: f64 = 0.0099;
const TRANSACTION_FEE_POS_LITE_SUMUP_ONE: f64 = 0.0099;
const TRANSACTION_FEE_TAP_TO_PAY_IPHONE_SUMUP_ONE: f64 = 0.0099;
const TRANSACTION_FEE_REMOTE_PAYMENT_SUMUP_ONE: f64 = 0.0099;
const TRANSACTION_FEE_QR_CODE_SUMUP_ONE: f64 = 0.0;

#[derive(Debug)]
pub enum PaymentOption {
    CardReader,
    PosLite,
    TapToPayIPhone,
    RemotePayment,
    QrCode,
}

#[derive(Debug)]
pub enum SubscriptionOption {
    NoContract,
    SumUpOne,
}

pub struct SumUpSale {
    pub cost: f64,
    pub payment_option: PaymentOption,
    pub subscription_option: SubscriptionOption,
}

impl SumUpSale {
    fn get_payment_option(&self) -> &PaymentOption {
        &self.payment_option
    }

    fn get_subscription_option(&self) -> &SubscriptionOption {
        &self.subscription_option
    }
}

impl Sale for SumUpSale {
    fn get_cost(&self) -> f64 {
        self.cost
    }

    fn get_delivery_costs(&self) -> f64 {
        0.0
    }
}

pub struct SumUpCharge {
    pub number_of_minutes: f64,
    pub material_costs: Vec<Material>,
    pub payment_option: PaymentOption,
    pub subscription_option: SubscriptionOption,
}

impl SumUpCharge {
    fn get_payment_option(&self) -> &PaymentOption {
        &self.payment_option
    }

    fn get_subscription_option(&self) -> &SubscriptionOption {
        &self.subscription_option
    }
}

impl Charge for SumUpCharge {
    fn get_number_of_minutes(&self) -> f64 {
        self.number_of_minutes
    }

    fn get_material_costs(&self) -> &Vec<Material> {
        &self.material_costs
    }

    fn get_delivery_costs(&self) -> f64 {
        0.0
    }
}

pub struct SumUpCalculator {}

impl Calculator<SumUpSale, SumUpCharge> for SumUpCalculator {
    fn based_on_sale(&self, config: &dyn Config, sale: SumUpSale) -> SaleBreakdown {
        let transaction_fee = match sale.payment_option {
            PaymentOption::CardReader => {
                if matches!(sale.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_CARD_READER_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_CARD_READER_SUMUP_ONE
                }
            }
            PaymentOption::PosLite => {
                if matches!(sale.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_POS_LITE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_POS_LITE_SUMUP_ONE
                }
            }
            PaymentOption::TapToPayIPhone => {
                if matches!(sale.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_SUMUP_ONE
                }
            }
            PaymentOption::RemotePayment => {
                if matches!(sale.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_REMOTE_PAYMENT_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_REMOTE_PAYMENT_SUMUP_ONE
                }
            }
            PaymentOption::QrCode => {
                if matches!(sale.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_QR_CODE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_QR_CODE_SUMUP_ONE
                }
            }
        };

        let transaction_cost = sale.cost * transaction_fee;

        let tax = sale.cost * (config.get_tax_rate() / 100.0);

        let revenue = sale.cost - transaction_cost - tax;
        let percentage_kept = (revenue / sale.cost) * 100.0;
        let max_working_hours = revenue / config.get_hourly_rate();

        SaleBreakdown {
            sale: sale.cost,
            delivery_costs: 0.0,
            transaction_cost,
            payment_processing_cost: 0.0,
            offsite_ads_cost: 0.0,
            regulatory_operating_fee: 0.0,
            vat_paid_by_buyer: 0.0,
            vat_on_seller_fees: 0.0,
            total_fees: transaction_cost,
            total_fees_with_vat: transaction_cost,
            tax,
            revenue,
            percentage_kept,
            max_working_hours,
        }
    }

    fn how_much_to_charge(&self, config: &dyn Config, charge: SumUpCharge) -> ChargeAmount {
        let total_material_costs: f64 = charge
            .material_costs
            .iter()
            .map(|material| material.value)
            .sum();
        let base_charge =
            ((charge.number_of_minutes / 60.0) * config.get_hourly_rate()) + total_material_costs;

        let transaction_fee = match charge.payment_option {
            PaymentOption::CardReader => {
                if matches!(charge.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_CARD_READER_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_CARD_READER_SUMUP_ONE
                }
            }
            PaymentOption::PosLite => {
                if matches!(charge.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_POS_LITE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_POS_LITE_SUMUP_ONE
                }
            }
            PaymentOption::TapToPayIPhone => {
                if matches!(charge.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_SUMUP_ONE
                }
            }
            PaymentOption::RemotePayment => {
                if matches!(charge.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_REMOTE_PAYMENT_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_REMOTE_PAYMENT_SUMUP_ONE
                }
            }
            PaymentOption::QrCode => {
                if matches!(charge.subscription_option, SubscriptionOption::NoContract) {
                    TRANSACTION_FEE_QR_CODE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_QR_CODE_SUMUP_ONE
                }
            }
        };

        let transaction_cost = base_charge * transaction_fee;
        let charge = base_charge + transaction_cost;

        let markup_percentage = (100.0 + config.get_markup_percentage()) / 100.0;
        let total_to_charge = (charge * markup_percentage).ceil();
        let with_vat = total_to_charge * ((config.get_vat() / 100.0) + 1.0);

        ChargeAmount {
            total_to_charge,
            with_vat,
        }
    }
}

File: composeApp/src/commonMain/kotlin/com/benjaminsproule/costcalculator/calculator/SumUpCalculator.kt
package com.benjaminsproule.costcalculator.calculator

import com.benjaminsproule.costcalculator.store.Config
import kotlin.math.ceil

interface Option {
    val label: String
}

enum class PaymentOption(override val label: String) : Option {
    CardReader("Card reader"),
    PosLite("POS lite"),
    TapToPayIPhone("Tap to pay iPhone"),
    RemotePayment("Remote payment"),
    QrCode("QR code"),
}

enum class SubscriptionOption(override val label: String) : Option {
    NoContract("No contract"),
    SumUpOne("SumUp One"),
}

data class SumUpSale(
    override val cost: Float,
    val paymentOption: PaymentOption,
    val subscriptionOption: SubscriptionOption
) : Sale() {
    override val deliveryCosts: Float
        get() = 0.0f
}

data class SumUpCharge(
    override val numberOfMinutes: Float,
    override val materialCosts: List<Material>,
    val paymentOption: PaymentOption,
    val subscriptionOption: SubscriptionOption
) : Charge() {
    override val deliveryCosts: Float
        get() = 0.0f
}

class SumUpCalculator(private val config: Config) : Calculator<SumUpSale, SumUpCharge> {
    companion object {
        const val TRANSACTION_FEE_CARD_READER_NO_CONTRACT: Float = 0.0169f
        const val TRANSACTION_FEE_POS_LITE_NO_CONTRACT: Float = 0.0169f
        const val TRANSACTION_FEE_TAP_TO_PAY_IPHONE_NO_CONTRACT: Float = 0.0169f
        const val TRANSACTION_FEE_REMOTE_PAYMENT_NO_CONTRACT: Float = 0.025f
        const val TRANSACTION_FEE_QR_CODE_NO_CONTRACT: Float = 0.0f
        const val TRANSACTION_FEE_CARD_READER_SUMUP_ONE: Float = 0.0099f
        const val TRANSACTION_FEE_POS_LITE_SUMUP_ONE: Float = 0.0099f
        const val TRANSACTION_FEE_TAP_TO_PAY_IPHONE_SUMUP_ONE: Float = 0.0099f
        const val TRANSACTION_FEE_REMOTE_PAYMENT_SUMUP_ONE: Float = 0.0099f
        const val TRANSACTION_FEE_QR_CODE_SUMUP_ONE: Float = 0.0f
    }

    override fun basedOnSale(sale: SumUpSale): SaleBreakdown {
        val transactionFee = when (sale.paymentOption) {
            PaymentOption.CardReader -> {
                if (sale.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_CARD_READER_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_CARD_READER_SUMUP_ONE
                }
            }

            PaymentOption.PosLite -> {
                if (sale.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_POS_LITE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_POS_LITE_SUMUP_ONE
                }
            }

            PaymentOption.TapToPayIPhone -> {
                if (sale.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_SUMUP_ONE
                }
            }

            PaymentOption.RemotePayment -> {
                if (sale.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_REMOTE_PAYMENT_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_REMOTE_PAYMENT_SUMUP_ONE
                }
            }

            PaymentOption.QrCode -> {
                if (sale.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_QR_CODE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_QR_CODE_SUMUP_ONE
                }
            }
        }

        val transactionCost = sale.cost * transactionFee;

        val tax = sale.cost * (config.taxRate / 100.0f);

        val revenue = sale.cost - transactionCost - tax;
        val percentageKept = (revenue / sale.cost) * 100.0f;
        val maxWorkingHours = revenue / config.hourlyRate;

        return SaleBreakdown(
            sale = sale.cost,
            deliveryCosts = 0.0f,
            transactionCost = transactionCost,
            paymentProcessingCost = 0.0f,
            offsiteAdsCost = 0.0f,
            regulatoryOperatingFee = 0.0f,
            vatPaidByBuyer = 0.0f,
            vatOnSellerFees = 0.0f,
            totalFees = transactionCost,
            totalFeesWithVat = transactionCost,
            tax = tax,
            revenue = revenue,
            percentageKept = percentageKept,
            maxWorkingHours = maxWorkingHours,
        )
    }

    override fun howMuchToCharge(
        charge: SumUpCharge
    ): ChargeAmount {
        val totalMaterialCosts = charge.materialCosts.map { material -> material.value }.sum();
        val baseCharge = ((charge.numberOfMinutes / 60.0f) * config.hourlyRate) + totalMaterialCosts;

        val transactionFee = when (charge.paymentOption) {
            PaymentOption.CardReader -> {
                if (charge.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_CARD_READER_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_CARD_READER_SUMUP_ONE
                }
            }

            PaymentOption.PosLite -> {
                if (charge.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_POS_LITE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_POS_LITE_SUMUP_ONE
                }
            }

            PaymentOption.TapToPayIPhone -> {
                if (charge.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_TAP_TO_PAY_IPHONE_SUMUP_ONE
                }
            }

            PaymentOption.RemotePayment -> {
                if (charge.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_REMOTE_PAYMENT_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_REMOTE_PAYMENT_SUMUP_ONE
                }
            }

            PaymentOption.QrCode -> {
                if (charge.subscriptionOption == SubscriptionOption.NoContract) {
                    TRANSACTION_FEE_QR_CODE_NO_CONTRACT
                } else {
                    TRANSACTION_FEE_QR_CODE_SUMUP_ONE
                }
            }
        }

        val transactionCost = baseCharge * transactionFee;
        val chargeWithFees = baseCharge + transactionCost;

        val markupPercentage = (100.0f + config.markupPercentage) / 100.0f;
        val totalToCharge = ceil(chargeWithFees * markupPercentage);
        val withVat = totalToCharge * ((config.vat / 100.0f) + 1.0f);

        return ChargeAmount(
            totalToCharge = totalToCharge,
            withVat = withVat,
        )
    }
}
